### ISSUE TITLE
Cut/paste across pages fails for vast majority of widgets

### ISSUE THREAD

**Issue by billreed63** [maqetta/maqetta#2367]
NEW SUMMARY by @aerwin:
In short, I tried a sampling of widgets where I'd cut from one page and try and paste into another. I found the following work:
- Dojo Button
- HTML widgets

Other than that, I couldn't find anything else.

Representative scenario:

New HTML file1
New HTML file2
Drag widget (say, Dojo TextBox) onto file1 canvas
Cut the TextBox
Switch to file2
Issue past command

~~OLD SUMMARY~~:
~~1) new file1.html~~
~~2)new file2.html~~
~~3) DnD Gridx on file1.html~~
~~4) Select GridX cut~~
~~5) Select file2.html~~
~~6) Paste, I get the cross hairs  but when I select on the canvas nothing happends~~

~~Note that when you cut the GridX widget from file1.html the itemStore is left behind, for all cut/paste/delete/move operation the item store has to move with the grid widget to ensure the store is defined before the grid~~

**Comment by aerwin** [comment]
@billreed63 I can reproduce, but this also doesn't appear to work with DataGrid. Most of the data store management for GridX is the code that was used for DataGrid. Did this used to work with DataGrid?

**Comment by jhpedemonte** [comment]
I think we discussed this several months back -- we shouldn't depend on positioning for these dependent widgets and having to move them around.

The easiest approach is just to put the stores as the first items in the **body** and make sure that all other (visible) widgets come after the stores in the DOM & Model.

**Comment by aerwin** [comment]
Still investigating, but this scenario appears to die on line 142 of `DataGridCreateTool` (with `storeWidget` being undefined):

```
var storeData = storeWidget.getData();
```

**Comment by billreed63** [comment]
@JonFerraiolo  wanted the Stores immediate before the widget that used it, I do not know if this is still his prefernnce but it was for DataGrid. This may be desirable if the GridX is a child of a div and the user cuts and pastes the parent div. if the store is a child of the div it would make cutting and pasting easier. Also more error proof in the source editor..

**Comment by aerwin** [comment]
Per scrum, pushing this to M7...

@jhpedemonte raised concern that we don't have any sort of test plan for pasting any widgets across pages (nor has that been discussed as a valid scenario). In quick testing, most seem to work, but @billreed63 has now opened an issue to create a test plan (#2370).

In the case of DataGrid/GridX, it looks to me that the current code in `DataGridCreateTool` is trying to get the store widget using `Widget.byId(storeId);` but this would be trying to look up the widget in the newly created page (where it doesn't exist) instead of the first page that the widget was copied from. So, I'm not sure this ever could have worked in the cross page scenario.

```
addPasteCreateCommand: function(command, args) {
        this._context = this._data.context;
        var storeId = this._data.properties.store._edit_object_id;
        //var storeId = this._widget._srcElement.getAttribute("store"); 
        var storeWidget = Widget.byId(storeId);
        var storeData = storeWidget.getData();
        var data = [];
        data[0] = storeData;
        data[1] = this._data;
        this._data = data;
        command.add( this._getCreateCommand(args));
        return this._dataGrid;
    }
```

**Comment by JonFerraiolo** [comment]
> @JonFerraiolo wanted the Stores immediate before the widget that used it, I do not know if this is still his prefernnce but it was for DataGrid. 

Yes it is still my preference, but obviously we are very late in the M6 cycle. I am seeing that users do a lot of source viewing and editing, so it's better to leave the data stores right in front the of the widgets that use the data stores.

If we can switch to dojo.mvc, this all might become moot, at least in some cases.

**Comment by aerwin** [comment]
-> Changing title of issue based on discussion below

In looking at this, I don't think there's anything GridX specific going on here. Or, if there are grid specific problems, more global problems need to be solved first. 

In short, I tried a sampling of widgets where I'd cut from one page and try and paste into another. I found the following work:
- Dojo Button
- HTML widgets

Other than that, I couldn't find anything else. 

Representative scenario:
1. New HTML file1
2. New HTML file2
3. Drag widget (say, Dojo TextBox) onto file1 canvas
4. Cut the TextBox
5. Switch to file2
6. Issue past command

There are no exceptions in the console, but turning on stop on all errors shows stops on line 168 of `dojo/parser`:

```
throw new Error("Unable to resolve constructor for: '" + obj.types.join() + "'");
```

`obj.types.join()` is properly set to `dijit.form.TextBox`.

What I think are the most relevant parts of the stack trace are bolded below:

(anonymous function) (parser.js:168)
array.map (array.js:278)
parser._instantiate (parser.js:164)
parser.instantiate (parser.js:138)
**declare.constructor (DijitWidget.js:50)**
(anonymous function) (declare.js:384)
**widgetObject.createWidget (widget.js:551)**
**declare._create (PasteTool.js:106)**
signalListener (Deferred.js:37)
Deferred.then.promise.then (Deferred.js:252)
declare._create (PasteTool.js:81)
ret.withDoc (window.js:121)
declare._create (PasteTool.js:79)
signalListener (Deferred.js:37)
Deferred.then.promise.then (Deferred.js:252)
declare._create (PasteTool.js:58)
declare._create (PasteTool.js:45)
declare.create (CreateTool.js:579)
signalListener (Deferred.js:37)
Deferred.then.promise.then (Deferred.js:252)
declare.create (CreateTool.js:578)
declare.onMouseUp (CreateTool.js:385)
declare.onMouseUp (Context.js:2287)
lang.hitch (lang.js:373)
